src/hotspot/share/opto/memnode.hpp

```diff
@@ -574,6 +574,10 @@ class StoreNode : public MemNode {
   Node *Ideal_masked_input       (PhaseGVN *phase, uint mask);
   Node *Ideal_sign_extended_input(PhaseGVN *phase, int  num_bits);
 
+  Node* Ideal_merge_primitive_array_stores(PhaseGVN* phase);
+  StoreNode* can_merge_primitive_array_store_with_use(PhaseGVN* phase, bool check_def);
```

Comment:
If I understand correctly you need the `check_def ` parameter to avoid having `can_merge_primitive_array_store_with_use` and `can_merge_primitive_array_store_with_def` call each other indefinitely. But if I was to write new code that takes advantage of one of the two methods, I think I would be puzzled that there's a `check_def` parameter. Passing `false` would be wrong then but maybe not immediately obvious. Maybe it would be better to have `can_merge_primitive_array_store_with_def` with no `check_def` parameter and have all the work done in a utility method that takes a `check_def` parameter (always `true` when called from `can_merge_primitive_array_store_with_def`)

Reply:
You are right, this is not the best code pattern. I'll refactor it.